pkg/agent/controller/packetsampling/packetsampling_controller.go

```diff
@@ -0,0 +1,759 @@
+		}
+
+	}
+	return nil
+}
+
+func (c *Controller) errorPacketSamplingCRD(ps *crdv1alpha1.PacketSampling, reason string) (*crdv1alpha1.PacketSampling, error) {
+	ps.Status.Phase = crdv1alpha1.PacketSamplingFailed
+
+	type PacketSampling struct {
+		Status crdv1alpha1.PacketSamplingStatus `json:"status,omitempty"`
+	}
+
+	patchData := PacketSampling{
+		Status: crdv1alpha1.PacketSamplingStatus{Phase: ps.Status.Phase, Reason: reason},
+	}
+	payloads, _ := json.Marshal(patchData)
+	return c.crdClient.CrdV1alpha1().PacketSamplings().Patch(context.TODO(), ps.Name, types.MergePatchType, payloads, metav1.PatchOptions{}, "status")
+
+}
+
+func (c *Controller) cleanupPacketSampling(psName string) {
+	psState := c.deletePacketSamplingState(psName)
+	if psState != nil {
+		err := c.ofClient.UninstallPacketSamplingFlows(psState.tag)
+		if err != nil {
+			klog.Errorf("Error cleaning up flows for PacketSampling %s: %v", psName, err)
+		}
+
+		err = psState.pcapngWriter.Flush()
+		if err != nil {
+			klog.Errorf("Error flushing pcapng file for PacketSampling %s: %v", psName, err)
+		}
+		err = psState.pcapngFile.Close()
+		if err != nil {
+			klog.Errorf("Error closing pcapng file for PacketSampling %s: %v", psName, err)
+		}
+	}
+}
+
+func (c *Controller) deletePacketSamplingState(psName string) *packetSamplingState {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	for tag, state := range c.runningPacketSamplings {
+		if state.name == psName {
+			delete(c.runningPacketSamplings, tag)
+			return state
+		}
+	}
+	return nil
+}
+
+func (c *Controller) startPacketSampling(ps *crdv1alpha1.PacketSampling) error {
+	err := c.validatePacketSampling(ps)
+
+	func() {
+		if err != nil {
+			c.cleanupPacketSampling(ps.Name)
+			c.errorPacketSamplingCRD(ps, fmt.Sprintf("Node: %s, error:%+v", c.nodeConfig.Name, err))
+
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+
+	receiverOnly := false
+	senderOnly := true
+	var pod, ns string
+	if ps.Spec.Source.Pod != "" {
+		pod = ps.Spec.Source.Pod
+		ns = ps.Spec.Source.Namespace
+	} else {
+		receiverOnly = true
+		senderOnly = false
+		pod = ps.Spec.Destination.Pod
+		ns = ps.Spec.Destination.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	isSender := !receiverOnly && len(podInterfaces) > 0
+
+	var matchPacket *binding.Packet
+	var ofPort uint32
+
+	if len(podInterfaces) > 0 {
+		packet, err := c.preparePacket(ps, podInterfaces[0], receiverOnly)
+		if err != nil {
+			return err
+		}
+		ofPort = uint32(podInterfaces[0].OFPort)
+		matchPacket = packet
+		klog.V(2).Infof("PacketSampling packet: %v", *packet)
+	}
+
+	c.runningPacketSamplingsMutex.Lock()
+	psState := packetSamplingState{
+		name: ps.Name, tag: uint8(ps.Status.DataplaneTag),
+		receiverOnly: receiverOnly, isSender: isSender,
+		maxNumCapturedPackets: ps.Spec.FirstNSamplingConfig.Number,
+	}
+
+	exists, err := fileExists(ps.Status.UID)
```

Comment:
can you call uidtopath before this line, so you can pass name string to `fileExists` and `createPcapngFile` instead of uid? so you don't have to call `uidtopath` twice.
Btw, I feel `createPcapngFile` is unnecessary if it's used only once.

Reply:
updated.